File: JavaAndScala/src/main/java/com/java/se/conclusion/design/observer/subjects/WeatherStation.java
package com.java.se.conclusion.design.observer.subjects;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import com.java.se.conclusion.design.observer.notifiers.WeatherNotifier;

/**
 * 	This is a class to implement observer pattern
 *  -- This class serves as a subject (that needs to be observed)
 *  
 *  To implement the observer pattern:
 * 	-- Offer a notifier interface for notification 
 *  -- Maintain a private reference list of all observers that implements the offered interface: IOP and polymorphism
 * 	-- Offer a public add and a public delete method to add and delete observers respectively
 *  -- Notify all observers actively through notifiers if there is any change of the subject
 *  
 * @author VinceYuan
 *
 */
public class WeatherStation {

	/*	Instance variables	*/
	private String[] weathers = {"Sunny", "Hazy", "Windy", "Hailing", "Snowy"};
	public String currentWeather;
	
	/**
	 * 	This is a reference list of all observers that implements the offered interface: IOP and polymorphism
	 */
	private List<WeatherNotifier> weatherNotifiers = new ArrayList<>();
	
	/**
	 * 	This is a method to add an observer
	 * @param weatherNotifier
	 */
	public void addObservers(WeatherNotifier weatherNotifier) {
		weatherNotifiers.add(weatherNotifier);
	}
	
	/**
	 * 	This is a method to remove an observer
	 * @param weatherNotifier
	 */
	public void deleteObservers(WeatherNotifier weatherNotifier) {
		weatherNotifiers.remove(weatherNotifier);
	}
	
	/*	This is a method to start the weather station	*/
	public void startWeatherStation() {
		
		/*	Loop to make sure weather station is always working	 */
		while (true) {
			try {
				/*	Update the current weather randomly	 */
				currentWeather = weathers[(int) (Math.random() * weathers.length)];
				System.out.println("Current weather: " + currentWeather);
				
				/**
				 * 	Notify all observers actively through notifiers if there is any change of the subject
				 * 	-- Here is actually a method polymorphism (or equivalent to callback in JavaScript), hence 1st way is equivalent to 2nd way
				 */
				for (WeatherNotifier weatherNotifier : weatherNotifiers) {
//					weatherNotifier.takeActionAccordingToCurrentWeather(currentWeather);	// 1st way
					takeActionAccordingToCurrentWeather(currentWeather, weatherNotifier);	// 2nd way
				}
				
				/*	Wait 1 to 1.5 seconds randomly	*/
				Thread.sleep(new Random().nextInt(501) + 1000);
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
	}
	
	/**
	 * 	This is a (optional) callback method
	 * 	-- The callback here is just for the demonstration of use of callback
	 *  -- Practically this is unnecessary
	 *  
	 * @param currentWeather
	 * @param weatherNotifier
	 */
	private void takeActionAccordingToCurrentWeather(String currentWeather, WeatherNotifier weatherNotifier) {
		weatherNotifier.takeActionAccordingToCurrentWeather(currentWeather);
	}
}

File: JavaAndScala/src/main/scala/org/lang/scala/conclusion/design/observer/subject/WeatherStation.scala
package org.lang.scala.conclusion.design.observer.subject

import scala.collection.mutable.ListBuffer
import scala.util.Random

import org.lang.scala.conclusion.design.observer.notifier.WeatherNotifier

/**
 * 	This is a stand-alone class to implement observer pattern
 *  -- This class serves as a subject (that needs to be observed)
 *  
 *  To implement observer pattern
 *  -- Need to maintain a data structure to store all observers (which extend notifiers)
 *     -- The data structure can be List, Array, ListBuffer, etc.
 *     -- Need to provide necessary methods to add or remove observers
 *  -- Need to notify all observers whenever need to notify them
 *     -- E.g., if there is any change of the subject
 *  
 * 	@author VinceYuan
 */
class WeatherStation {
  
  /*	instance variables that provides all possible weather	*/
  private val weathers = Array("Sunny", "Hazy", "Windy", "Hailing", "Snowy")
  
  /**
   * 	instance variable	that contain all observers that need to observe this subject
   */
  private val observers = ListBuffer[WeatherNotifier]()
  
  /**
   * 	Here are the methods to add and remove an observer
   */
  def addObserver(observer: WeatherNotifier): Unit = {
    observers += observer
  }
  def removeObserver(observer: WeatherNotifier): Unit = {
    observers -= observer
  }
  
  /**
   * 	This is a method to start the weather station
   */
  def startWorking() = {
    
    /*	Loop forever	*/
    while (true) {
      
      /*	Get current weather	 */
      val currentWeather = weathers(Random.nextInt(weathers.length))
      println(s"Current weather: ${currentWeather}")
      
      /**	
       * 	Notify each observer
       *  -- Since each observe need to know (the change of) current weather
       */
      observers.foreach(observer => {
        observer.takeActionAccordingToCurrentWeather(currentWeather)
      })
      
      /*	Current thread sleeps for 1 to 1.5 second	 */
      Thread.sleep(1000 + Random.nextInt(501))
    }
  }
}